Add validation for empty fields and duplicate nicknames in user updat…

agusbusta opened the pull request:
## User Update Endpoint Validation Enhancement

### Changes Made
- Added validation to prevent empty field updates in user data
- Implemented nickname uniqueness validation
- Added check to prevent updating nickname to its current value

### Technical Details
- Added string validation to ensure non-empty values for text fields
- Enhanced nickname validation to check against existing users
- Added appropriate error responses (400, 409) with descriptive messages

### Testing
The following scenarios have been tested:
- Empty field submission
- Duplicate nickname attempts
- Self-duplicate nickname attempts
- Valid data updates

### Expected Behavior
- Returns 400 if fields are empty
- Returns 409 if nickname already exists
- Returns 400 if trying to update nickname to current value
- Returns 200 with updated data for valid requests

agusbusta commented:
## Changes
- Added new Auth0 management functions:
  - `deleteUserAuth0`: Handles user deletion in Auth0
  - `updateUserAuth0`: Manages user data updates in Auth0
- Modified user endpoints:
  - Made `/user/<user_id>` PUT endpoint async to handle Auth0 updates
  - Updated `/user` DELETE endpoint to remove users from both Auth0 and local DB
- Added error handling for Auth0 API operations
- Converted relevant endpoints to async/await pattern

## Dependencies
- Requires Auth0 Management API credentials in `.env`:
  - `AUTH0_DOMAIN`
  - `AUTH0_MANAGEMENTAPI_CLIENT`
  - `AUTH0_MANAGEMENTAPI_SECRET`

## Security Considerations
- All Auth0 API calls use secure token-based authentication
- Sensitive operations are properly error-handled
- User data synchronization maintains data integrity

daviddflix commented:
Hi @agusbusta, 

I’ve reviewed the code, and while it looks fine overall, I encountered an error while testing the `PUT` endpoint. The error seems to be related to a coroutine issue, which is often caused by improperly handled async functions. Below is the relevant error message for reference:

```python
TypeError: The view function did not return a valid response. The return type must be a string, dict, list, tuple with headers or status, Response instance, or WSGI callable, but it was a coroutine. // Werkzeug Debugger
```

A few observations and recommendations:
1. I noticed that we have `flask[async]` in the `requirements.txt`, but it’s likely not being utilized properly.
2. I couldn’t test this in the Docker environment because the `docker-compose.dev.yml` file has been upgraded. I recommend updating this file to match the current development version. 

### Next Steps:
1. Update the `docker-compose.dev.yml` file to the latest version for development.
2. Test the endpoint in a Docker environment.
3. Provide evidence of the testing using the Postman instance we have available.

Additionally, we need to clean up the `auth.py` file located at `routes/user/auth.py`. Some functions are duplicated. Please retain only the following functions:
- `get_management_api_token`
- `get_users_by_name_or_email`
- `patchPassword`
- `list_all_users`
- `deleteUserAuth0`
- `updateUserAuth0`

Let me know if you have any questions or need further clarification. Thanks!

agusbusta commented:
## Key Changes

### Auth0 Synchronization
- Added async support for Auth0 operations in user management endpoints
- Implemented smart handling of different Auth0 connection types (database vs social)
- Added graceful error handling for Auth0 operations
- User updates now sync properly between local DB and Auth0
- User deletions are now properly cascaded to Auth0

### Code Improvements
- Refactored auth.py to remove duplicate functions
- Organized authentication-related functions into clear, single-responsibility methods
- Enhanced error handling and logging
- Improved function documentation and type hints
- Standardized API responses

### Modified Files
- `routes/user/auth.py`: Cleaned up and reorganized Auth0 integration functions
- `routes/user/user.py`: Added async support and Auth0 sync for PUT/DELETE operations

## Technical Details

### New Auth0 Integration Features
- Smart field updates based on connection type:
  - Database connections: Can update nickname and full name
  - Social connections (Apple, Google): Only update name field
- Proper error handling for Auth0 API restrictions
- Automatic fallback to local updates when Auth0 restrictions apply

### Async Implementation
- Added async/await pattern for Auth0 operations
- Implemented proper async route handling in Flask
- Added error handling specific to async operations

## Testing Instructions
1. Test user updates with different Auth0 connection types:
   - Database users
   - Social login users (Apple, Google)
2. Verify user deletion cascade:
   - Check local database cleanup
   - Verify Auth0 user removal
3. Test error scenarios:
   - Network issues
   - Auth0 API restrictions
   - Invalid data
